checker/src/main/java/org/checkerframework/checker/mustcall/MustCallAnnotatedTypeFactory.java

```diff
@@ -220,6 +225,78 @@ protected void constructorFromUsePreSubstitution(
     super.constructorFromUsePreSubstitution(tree, type);
   }
 
+  /**
+   * Class to implement the customized semantics of {@link MustCallAlias} (and {@link PolyMustCall})
+   * annotations; see the {@link MustCallAlias} documentation for details.
+   */
+  private class MustCallQualifierPolymorphism extends DefaultQualifierPolymorphism {
+    /**
+     * Creates a {@link MustCallQualifierPolymorphism}.
+     *
+     * @param env the processing environment
+     * @param factory the factory for the current checker
+     */
+    public MustCallQualifierPolymorphism(ProcessingEnvironment env, AnnotatedTypeFactory factory) {
+      super(env, factory);
+    }
+
+    @Override
+    protected void replace(
+        AnnotatedTypeMirror type, AnnotationMirrorMap<AnnotationMirror> replacements) {
+      boolean replacementsModified = false;
+      AnnotationMirror extantPolyAnnoReplacement = null;
+      TypeElement typeElement = TypesUtils.getTypeElement(type.getUnderlyingType());
+      // only customize replacement for type elements
+      if (typeElement != null) {
+        // TODO: should we assert that replacements.size() == 1?
+        if (replacements.size() == 1 && replacements.containsKey(POLY)) {
+          extantPolyAnnoReplacement = replacements.get(POLY);
+          if (AnnotationUtils.areSameByName(
+              extantPolyAnnoReplacement, MustCall.class.getCanonicalName())) {
+            List<String> extentReplacementVals =
+                AnnotationUtils.getElementValueArray(
+                    extantPolyAnnoReplacement,
+                    getMustCallValueElement(),
+                    String.class,
+                    Collections.emptyList());
+            // replacement only customized when parameter type has a non-empty must-call obligation
+            if (!extentReplacementVals.isEmpty()) {
+              AnnotationMirror inheritableMustCall =
+                  getDeclAnnotation(typeElement, InheritableMustCall.class);
+              if (inheritableMustCall != null) {
+                List<String> inheritableMustCallVals =
+                    AnnotationUtils.getElementValueArray(
+                        inheritableMustCall,
+                        inheritableMustCallValueElement,
+                        String.class,
+                        Collections.emptyList());
+                if (!inheritableMustCallVals.equals(extentReplacementVals)) {
+                  // Use the must call values from the @InheritableMustCall annotation instead.
+                  // This allows for wrapper types to have a must-call method with a different
+                  // name than the must-call method for the wrapped type
+                  // TODO should we just copy the map instead of modifying it?
```

Comment:
If we know that `replacements` will always have a single entry, I will get rid of this mutation, and instead just create an alternate single-entry `ImmutableMap`.

Reply:
That sounds good.